packages/discount-code-exporter/test/main.spec.js

```diff
@@ -57,70 +57,146 @@ describe('DiscountCodeExport', () => {
-          `
-        expect(result).toEqual(expectedResult)
-        done()
+  describe('::setupHeaders', () => {
+    describe('with passed headers', () => {
+      test('should return return passed headers', () => {
+        expect(DiscountCodeExport.setupHeaders(['foo', 'bar'], 'de')).toEqual([
+          'foo',
+          'bar',
+        ])
       })
-      codeExport.run(outputStream)
     })
 
-    test('should fetch codes and output json to stream by default', done => {
-      const sampleCode = {
-        code: 'discount-code',
-        name: { en: 'some-discount-name' },
-        cartDiscounts: [{ id: 'cart-discount-1' }, { id: 'cart-discount-2' }],
-      }
-      codeExport._fetchCodes = jest.fn().mockImplementation(jsonStream => {
-        jsonStream.write(sampleCode)
-        return Promise.resolve()
+    describe('without passed headers', () => {
+      describe('when empty array', () => {
+        test('should return return passed headers', () => {
+          expect(DiscountCodeExport.setupHeaders([], 'de')).toMatchSnapshot()
+        })
       })
-      const outputStream = streamtest.v2.toText((error, result) => {
-        const expectedResult = [sampleCode]
-        expect(JSON.parse(result)).toEqual(expectedResult)
-        done()
+
+      describe('when null', () => {
+        test('should return return passed headers', () => {
+          expect(DiscountCodeExport.setupHeaders(null, 'de')).toMatchSnapshot()
+        })
       })
-      codeExport.run(outputStream)
     })
+  })
+
+  describe('::run', () => {
+    describe('successful export', () => {
+      let sampleCode
+      beforeEach(() => {
+        sampleCode = {
+          code: 'discount-code',
+          name: { en: 'some-discount-name' },
+          cartDiscounts: [{ id: 'cart-discount-1' }, { id: 'cart-discount-2' }],
+        }
+        codeExport._fetchCodes = jest.fn().mockImplementation(csvStream => {
+          csvStream.write(sampleCode)
+          return Promise.resolve()
+        })
+      })
 
-    test('should emit error if it occurs when streaming to csv', done => {
-      codeExport.exportFormat = 'csv'
-      codeExport._fetchCodes = jest
-        .fn()
-        .mockImplementation(() => Promise.reject(new Error('error occured')))
-      const outputStream = streamtest.v2.toText((error, result) => {
-        expect(error.message).toBe('error occured')
-        expect(result).toBeUndefined()
-        done()
+      describe('CSV export', () => {
+        describe('with template', () => {
+          beforeEach(() => {
+            codeExport = new DiscountCodeExport(
+              {
+                apiConfig: {
+                  projectKey: 'test-project-key',
+                },
+                exportFormat: 'csv',
+                headerFields: ['code', 'name.en', 'cartDiscounts', 'groups'],
+              },
+              logger
+            )
+            codeExport._fetchCodes = jest.fn().mockImplementation(csvStream => {
```

Comment:
I know we talked about this and it's unrelated to this PR but it feels more and more that the `fetchCodes` should be a module in itself. It would make testing way easier. The `DiscountCodeExport` could just `import { fetchCodes } from './fetch-codes'` but the test would just `jest.mock('./fetch-codes')` while also `import { fetchCodes } from './fetch-codes'`. It can then just use `fetchCodes.mockImplementation`.

Reply:
True. I see a lot of sense in that. But like you rightly said, this isn't the PR for it. Nevertheless I think it's something that should be done at some point